Blog about javascript SDKs (#2595)
FusionAuth/fusionauth-site: astro/src/content/blog/javascript-sdks.mdx

---
publish_date: 2023-10-17 
title: Our New Vue, React and Angular SDKs
description: JavaScript SDKs make integrating with FusionAuth easier
authors: Dan Moore
excerpt_separator: "{/* more */}"
image: /img/blogs/hackfest/hackfest-fusionauth.png
categories: Company
tags: javascript, sdks, angular, vue, react, client libraries
---
import { YouTube } from '@astro-community/astro-embed-youtube';
import RemoteCode from "../../components/RemoteCode.astro";

FusionAuth is, at its core, an architectural component of a larger application. I like to say "no one ever builds a full application with just FusionAuth". Sure, your application needs authentication and authorization, but there are other critical components. For example, the features of your application which make users *want* to register and log in.

{/* more */}

But here at FusionAuth, we want simple, secure authentication to be easy to add to any application. FusionAuth can run anywhere, but what use is that if it isn't easy to integrate? Applications which FusionAuth works with include:

> * JavaScript single pages applications (SPAs)

Reviewer: ```suggestion
* JavaScript single page applications (SPAs)
```

* traditional webapps
* commercial off-the-shelf applications
* server-side applications
* APIs
* mobile apps

Pretty much anything which can use OpenID Connect (OIDC), Security Assertion Markup Language (SAML), or an HTTP based API should work. You can see our [list of quickstarts](/docs/quickstarts/) for integration examples.

Recently, the team built software development kits (SDKs) to make it even easier for SPA developers. The SDKs support the following JavaScript front-end frameworks:

* [Angular](/docs/v1/tech/client-libraries/angular-sdk)
* [React](/docs/v1/tech/client-libraries/react-sdk)
* [Vue](/docs/v1/tech/client-libraries/vue-sdk)

This post will cover the nitty gritty of using these SDKs, but first, let's discuss the difference between an SDK and a [client library](/docs/v1/tech/client-libraries/).

## Client Libraries vs SDKs

At FusionAuth, a client library is different from an SDK, even though many people in the industry use the terms interchangeably.

### Client Libraries

> A client library is a wrapper over [FusionAuth APIs](/docs/v1/tech/apis/). Client libraries support all public FusionAuth API endpoints and offer a convenient way to use that API in many software languages. At the time of writing, there are eight supported languages, including TypeScript. There's also an OpenAPI definition, useful for viewing API changes over time or generating a client library for an unsupported language like C++ or Rust.

Reviewer: ```suggestion
A client library is a wrapper over [FusionAuth APIs](/docs/v1/tech/apis/). Client libraries support all public FusionAuth API endpoints and offer a convenient way to use that API in many software languages. At the time of writing, there are eight supported languages, including TypeScript. There's also an OpenAPI definition, useful for viewing API changes over time or making HTTP calls from an unsupported language like C++ or Rust.
```

### The Client Side Package

The client-side package offers, as mentioned above, pre-built buttons or functions to which you can attach UI elements. Each of these performs a redirect to FusionAuth's hosted login pages for the corresponding operation; this includes the Authorization Code grant for login. 

This is the recommended approach for two reasons:

* more extensive functionality
* increased security

Let's discuss each of these in turn.

### Functionality

When you use the Authorization Code grant for authentication, as well as redirects for registration and logout, you are leveraging the FusionAuth provided functionality. Your application immediately supports a large number of authentication workflows. These include, but are not limited to:

* forgot password
* email verification
* [MFA](/docs/v1/tech/guides/multi-factor-authentication)
* [magic link authentication](/docs/v1/tech/passwordless/magic-links)
* password expiration
* [enterprise single sign-on (SAML, OIDC)](/docs/v1/tech/identity-providers/#enterprise-identity-providers)

> * [social sign-on (Google, Facebook, etc)](http://localhost:4000/docs/v1/tech/identity-providers/#social-identity-providers)

Reviewer: ```suggestion
* [social sign-on (Google, Facebook, etc)](/docs/v1/tech/identity-providers/#social-identity-providers)
```

categories: Company
tags: javascript, sdks, angular, vue, react, client libraries
---
import { YouTube } from '@astro-community/astro-embed-youtube';
import RemoteCode from "../../components/RemoteCode.astro";

FusionAuth is, at its core, an architectural component of a larger application. I like to say "no one ever builds a full application with just FusionAuth". Sure, your application needs authentication and authorization, but there are other critical components. For example, the features of your application which make users *want* to register and log in.

{/* more */}

But here at FusionAuth, we want simple, secure authentication to be easy to add to any application. FusionAuth can run anywhere, but what use is that if it isn't easy to integrate? Applications which FusionAuth works with include:

* JavaScript single pages applications (SPAs)
* traditional webapps
* commercial off the shelf applications
* server side applications
* APIs
* mobile apps

Pretty much anything which can use OpenID Connect (OIDC), Security Assertion Markup Language (SAML), or an HTTP based API should work. You can see our [list of quickstarts](/docs/quickstarts/) for integration examples.

> Recently, the team built software development kits (SDKs) to make it even easier for SPA developers. The SDKs support the following JavaScript front end frameworks:

Reviewer 2: ```suggestion
Recently, the team built software development kits (SDKs) to make it even easier for SPA developers. The SDKs support the following JavaScript front-end frameworks:
```

* [Angular](/docs/v1/tech/client-libraries/angular-sdk)
* [React](/docs/v1/tech/client-libraries/react-sdk)
* [Vue](/docs/v1/tech/client-libraries/vue-sdk)

This post will cover the nitty gritty of using these SDKS, but first, let's discuss the difference between an SDK and a [client library](/docs/v1/tech/client-libraries/).

## Client Libraries vs SDKs

At FusionAuth, a client library is different from an SDK, even though many people in the industry use the terms interchangeably.

### Client Libraries

A client library is a wrapper over [FusionAuth APIs](/docs/v1/tech/apis/). Client libraries support all public FusionAuth API endpoints, and offer a convenient way to use that API in many software languages. At time of writing, there are eight supported languages, including TypeScript. There's also an OpenAPI definition, useful for viewing API changes over time or generating a client library for an unsupported language like C++ or Rust.

> If you want to compare a client library to a child's toy, it is like tinker toys. Engineers build a script or program using a client library to extend or configure FusionAuth. Some examples of tasks well suited to a client library:

Reviewer 2: I don't get the analogy here. Maybe add that like tinker toys you have the building blocks but not the complete toy. The millennial in me also wouldn't reference a tinker toy...but that is for another day.

Author: I could use legos as the example too?

For SPAs which make requests to APIs not on the same domain, the backend for frontend pattern is recommended. You should not put bearer tokens into local storage or session storage. Here's a video about the insecurity of storing tokens anyplace accessible to JavaScript running in the browser.

<YouTube id="2nVYLruX76M" />

You can read more about [OAuth token storage in this article](/articles/oauth/oauth-token-storage). Using a server side component gives you flexibility to store the token wherever makes sense.

### The Client Side Package

The client side package offers, as mentioned above, pre-built buttons or functions to which you can attach UI elements. Each of these performs a redirect to FusionAuth's hosted login pages for the corresponding operation; this includes the Authorization Code grant for login. 

This is the recommended approach for two reasons:

* more extensive functionality
* increased security

Let's discuss each of these in turn.

### Functionality

> When you use the Authorization Code grant for authentication, as well as redirects for registration and logout, you are leveraging FusionAuth provided functionality. Your application immediately supports a large number of authentication workflows. These include, but are not limited to:

Reviewer 2: ```suggestion
When you use the Authorization Code grant for authentication, as well as redirects for registration and logout, you are leveraging the FusionAuth provided functionality. Your application immediately supports a large number of authentication workflows. These include, but are not limited to:
```

Reviewer 2: Should we add backlinks to these?

Author: Probably. Good call.

> 

Author: Add links here.

* forgot password
* email verification
* MFA
* magic link authentication
* password expiration
* enterprise single sign-on (SAML, OIDC)
* social sign-on (Google, Facebook, etc)
* profile management (with a paid license)
* passkeys (with a paid license)
* and more

Now, it's not entirely effortless. You must configure FusionAuth, but all this functionality is present, tested, and maintained over time. For example, instead of having to figure out how to set up email verification, you configure a customizable email and enable it in FusionAuth, and it works. In addition, browser redirects work with any language, framework or browser. This ensures your application's login page is never a barrier to your users.

### Security

> When using a redirect, the only application which sees or holes sensitive user credentials is FusionAuth. Your application code never has access to these, which decreases the attack surface area.

Reviewer 2: ```suggestion
When using a redirect, the only application that sees or holes sensitive user credentials is FusionAuth. Your application code never has access to these, which decreases the attack surface area.
```

In addition, FusionAuth has built-in security features such as:

* the ability to lock user accounts
* password rules
* webhook notifications when a user has failed to login
* breached password notification (with a paid license)
* rate limiting of sensitive security related actions (with an Enterprise license)
* and more

[FusionAuth takes security seriously](/security); it's our full time job. By redirecting to FusionAuth, you don't have to worry about any other JavaScript in your application getting access to sensitive information or credentials.

### Look and Feel

Additional functionality and security are great, but you typically want your login pages to look like your application. Doesn't a browser redirect make that difficult?

> Well, no. FusionAuth allows complete control over the look and feel of the hosted login pages, including adding or removing HTML, CSS, images and JavaScript to the pages. Learn more about this functionality, which we call [themes](/docs/v1/tech/themes/), including an [example using Tailwind CSS](/docs/v1/tech/themes/tailwind).

Reviewer 2: ```suggestion
Well, no. FusionAuth allows complete control over the look and feel of the hosted login pages, including adding or removing HTML, CSS, images, and JavaScript to the pages. Learn more about this functionality, which we call [themes](/docs/v1/tech/themes/), including an [example using Tailwind CSS](/docs/v1/tech/themes/tailwind).
```

One other common objection to redirection is the ability to send a successfully authenticated user right back to where they started. It's common to allow users to log in from multiple states in your SPA, after all. Luckily, you can use the `state` parameter to deep link from any page in your SPA to FusionAuth and then send the user right back to that page after a successful login. Here's a video displaying that:

> <YouTube id="-vx5rdy-mvY" />

Reviewer 2: Let's have Brand/Design get a cover on this one too! I don't think we should show any videos without those.

Author: Great, will add a task.

Here's code from an [example application](https://github.com/FusionAuth/fusionauth-example-node-deeplink/) which extracts a URL from a `state` parameter and returns the user to that URL after a successful exchange at the token endpoint.

<RemoteCode url="https://raw.githubusercontent.com/FusionAuth/fusionauth-example-node-deeplink/main/routes/index.js" lang="javascript" tags="extractURLFromState" />

The takeaway from this code is that you can set the `state` parameter when redirecting the user to the login page, then extract a value from it and redirect back to the correct location.

### Redirects Don't Work, Help!

If your application architecture, use case, or design team rules out redirects, you can use the [Login API](/docs/v1/tech/apis/login) to directly pass user credentials to FusionAuth. 

You'll miss out on the functionality and security benefits outlined above, however.

> At the current time there's no SDK support for this approach, but FusionAuth users can and have built out their own user interface using the Login API.

Reviewer 2: Do we have any examples we could link to on this?

> At the current time there's no SDK support for this approach, but FusionAuth users can and have built out their own user interface using the Login API.

Author: No, it's not really recommended. I actually wasn't sure I should include it, but it is a supported use case.

## In Conclusion

> Using the Angular, React or Vue SDKs can accelerate your app's integration with FusionAuth.

Reviewer 2: ```suggestion
Using the Angular, React, or Vue SDKs can accelerate your app's integration with FusionAuth.
```

> By doing so, you can focus on the features you users actually care about while leveraging a world-class implementation of undifferentiated login workflows and security of an isolated authentication server.

Reviewer 2: ```suggestion
By doing so, you can focus on the features your users actually care about while leveraging a world-class implementation of undifferentiated login workflows and the security of an isolated authentication server.
```

---
publish_date: 2023-10-17 
title: Our New Vue, React and Angular SDKs
description: JavaScript SDKs make integrating with FusionAuth easier
authors: Dan Moore
excerpt_separator: "{/* more */}"
image: /img/blogs/javascript-sdks/our-new-vue-react-and-angular-sdks.png
categories: Company
tags: javascript, sdks, angular, vue, react, client libraries
---
import { YouTube } from '@astro-community/astro-embed-youtube';
import RemoteCode from "../../components/RemoteCode.astro";

> FusionAuth is, at its core, an architectural component of a larger application. I like to say "no one ever builds a full application with just FusionAuth". Sure, your application needs authentication and authorization, but there are other critical components. For example, the features of your application which make users *want* to register and log in.

Reviewer 2: ```suggestion
FusionAuth is, at its core, an architectural component of a larger application. I like to say "no one ever builds a full application with just FusionAuth". Sure, your application needs authentication and authorization, but there are other critical components. For example, the features of your application that make users *want* to register and log in.
```

### Client Libraries

A client library, in contrast, is a wrapper over [FusionAuth APIs](/docs/v1/tech/apis/). Client libraries support all public FusionAuth API endpoints and offer a convenient way to use that API in many software languages. At the time of writing, there are eight supported languages, including TypeScript. There's also an OpenAPI definition, useful for viewing API changes over time or generating a client library for a language without an official client library, like C++ or Rust.

If you want to compare a client library to a child's toy, it is like legos. Engineers build a script or program using a client library to extend or configure FusionAuth. Some examples of tasks well suited to a client library:

* [rotating client secrets or API keys](/docs/v1/tech/tutorials/key-rotation) to improve your security posture
* pulling user data every night for data warehouse ingestion
* creating a new FusionAuth tenant every time a client signs up for your app
* offering a unique, custom login or registration experience not supported by the [FusionAuth hosted login pages](/docs/v1/tech/core-concepts/integration-points#hosted-login-pages)

> Requests made with client libraries usually require an API key as well as a URL for the FusionAuth instance which the requests will be applied to.

Reviewer 2: ```suggestion
Requests made with client libraries usually require an API key as well as a URL for the FusionAuth instance to which the requests will be applied.
```

### Better Together?

> You can use both an SDK and a client library if your use case requires the front-end SDK-based solutions for easy integration and the management access provided by a client library. They are not mutually exclusive.

Reviewer 2: ```suggestion
You can use both an SDK and a client library if your use case requires front-end SDK-based solutions for easy integration and the management access provided by a client library. They are not mutually exclusive.
```

[FusionAuth takes security seriously](/security); it's our full-time job. By redirecting to FusionAuth, you don't have to worry about any other JavaScript in your application getting access to sensitive information or credentials.

### Look and Feel

Additional functionality and security are great, but you typically want your login pages to look like your application. Doesn't a browser redirect make that difficult?

Well, no. FusionAuth allows complete control over the look and feel of the hosted login pages, including adding or removing HTML, CSS, images, and JavaScript to the pages. Learn more about this functionality, which we call [themes](/docs/v1/tech/themes/), including an [example using Tailwind CSS](/docs/v1/tech/themes/tailwind).

One other common objection to redirection is the ability to send a successfully authenticated user right back to where they started. It's common to allow users to log in from multiple states in your SPA, after all. Luckily, you can use the `state` parameter to deep link from any page in your SPA to FusionAuth and then send the user right back to that page after a successful login. Here's a video displaying that:

<YouTube id="-vx5rdy-mvY" />

> Here's code from an [example application](https://github.com/FusionAuth/fusionauth-example-node-deeplink/) which extracts a URL from a `state` parameter and returns the user to that URL after a successful exchange at the token endpoint.

Reviewer 2: ```suggestion
Here's code from an [example application](https://github.com/FusionAuth/fusionauth-example-node-deeplink/) that extracts a URL from a `state` parameter and returns the user to that URL after a successful exchange at the token endpoint.
```

The takeaway from this code is that you can set the `state` parameter when redirecting the user to the login page, then extract a value from it and redirect back to the correct location.

### The Server Side Component

The server-side component has endpoints that correspond to the [hosted backend APIs](/docs/v1/tech/apis/hosted-backend). You can implement the server-side component in any other language if that's preferable for your application. The server-side code sets access tokens and refresh tokens as `HTTPOnly`, `Secure` cookies. This is a great storage choice for first-party API requests because such cookies are not vulnerable to JavaScript exfiltration. The cookies are presented on every request from the SPA to an API, as long as you set the [`credentials` property correctly](https://developer.mozilla.org/en-US/docs/Web/API/Request/credentials) and the API is in the same domain that the SPA is served from. For this approach to work, APIs must look for an access token in a cookie, though they can also fall back to the `Authorization` header. You can see examples of this behavior in our [API quickstarts](/docs/quickstarts/#api). 

> For SPAs which must request APIs that don't live on the same domain the SPA is served from, cookies won't work. In this situation, the backend for frontend pattern is recommended. **You should not put bearer tokens into local storage or session storage.** Here's a video about the insecurity of storing tokens anyplace accessible to JavaScript running in the browser.

Reviewer 2: ```suggestion
For SPAs that must request APIs that don't live on the same domain the SPA is served from, cookies won't work. In this situation, the backend for front-end pattern is recommended. **You should not put bearer tokens into local storage or session storage.** Here's a video about the insecurity of storing tokens anyplace accessible to JavaScript running in the browser.
```

Reviewer 2: Maybe this should even be "Back-end for Front-end" not sure on that one.
